Adding support for Torchscript - Models

shrinath-suresh opened the pull request:
## What changes are proposed in this pull request?

Adding support for torchscript models. 

## How is this patch tested?

Tested mlflow.pytorch - save_model, log_model and load_model

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [x] `language/r`: python APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

smurching commented on mlflow/pytorch/__init__.py:
```diff
@@ -297,7 +297,11 @@ def save_model(pytorch_model, path, conda_env=None, mlflow_model=None, code_path
     # Save pytorch model
     model_path = os.path.join(model_data_path, _SERIALIZED_TORCH_MODEL_FILE_NAME)
-    torch.save(pytorch_model, model_path, pickle_module=pickle_module, **kwargs)
+    if isinstance(pytorch_model, torch.jit.ScriptModule):
```
BTW, we'd actually decided in https://github.com/mlflow/mlflow/issues/2263 that it made more sense to add torchscript model persistence support via a separate model flavor. There are two issues with having `mlflow.pytorch.log_model` support logging `torchscript` models:

1. At least currently, `mlflow.pytorch.load_model` cannot load back `torch.jit.ScriptModule` models (since we call `torch.load` to load back the model)
2. If we did add support for loading back `torch.jit.ScriptModule`s, downstream user code would have to type-check whether the loaded model was an instance of `torch.jit.ScriptModule` or `torch.nn.Module` after calling `loaded_model = mlflow.pytorch.load_model(...)`

Thus we'd discussed that it'd actually be cleaner to have separate `mlflow.torchscript.log_model/load_model` APIs. Let me know if that makes sense, or if I'm missing something!

shrinath-suresh replied:
@chauhang Can you please confirm if we can add a separate flavor for torch scripted models. We remember discussing the same during the call, where you had suggested us not to create a separate flavor.

smurching replied:
If there's a good way to add unified pytorch/torchscript support that'd be awesome, but at least I'm not aware of an approach that satisfies the issues mentioned above^ in my previous comment. Note that even with a separate torchscript flavor we could consider having `mlflow.pytorch.log_model` additionally generate the torchscript flavor, i.e. call  `torchscript_model = torch.jit.trace(pytorch_model); mlflow.torchscript.log_model(torchscript_model)` on the input model, if it's not too expensive to trace a model via `torch.jit.trace`

chauhang replied:
We discussed this on the call and it is resolved. We are adding examples for the pyfunc as well

shrinath-suresh commented:
@chauhang @smurching we have added an example for logging torch scripted model and updated the same test files to run on both regular and scripted model. Please review and let us know your comments. If the changes are fine, we will raise an upstream request for the same.

harupy commented:
@chauhang Could you enable GitHub Actions on this repo so that we can verify the tests work properly?

chauhang commented on examples/pytorch/torchscript/mnist_torchscript.py:
```diff
@@ -0,0 +1,250 @@
+    dict_args = vars(args)
+    mlflow.set_tracking_uri(dict_args["tracking_uri"])
+    model = LightningMNISTClassifier(**dict_args)
+    scripted_model = torch.jit.script(model)
```
Shouldn't this scripted model be saved after the model has been trained? The sequence of steps in the example is not clear

chauhang reviewed (commented):
Please see inline comments

harupy commented on examples/pytorch/torchscript/mnist_torchscript.py:
```diff
@@ -0,0 +1,186 @@
+    transform = transforms.Compose(
+        [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
+    )
+    dataset1 = datasets.MNIST("../data", train=True, download=True, transform=transform)
+    dataset2 = datasets.MNIST("../data", train=False, transform=transform)
```
I think training a model using the full MNIST dataset on a laptop without GPUs takes really long time. Could we use a smaller dataset (e.g. the iris dataset) instead?

shrinath-suresh replied:
We tested with lesser epoch .. didnt take much time to complete. We will add iris example as well

shrinath-suresh commented:
raised an upstream PR for this change - https://github.com/mlflow/mlflow/pull/3557

shrinath-suresh commented:
Merged in upstream. Closing it as duplicate
